opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -1369,7 +1369,12 @@ namespace Opm
 
         const int max_iter_number = this->param_.max_inner_iter_ms_wells_;
         const WellState well_state0 = well_state;
-        const std::vector<Scalar> residuals0 = this->getWellResiduals(Base::B_avg_, deferred_logger);
+        std::vector<Scalar> residuals(numWellEq + 1, 0.0);
+
+        // getWellResiduals returns true for nan/inf residuals
+        if(this->getWellResiduals(Base::B_avg_, residuals, deferred_logger))
+            return false;
```

Comment:
I know why it is this way, especially with the comment, but I'll still remark that
```C++
if (this->getX(...)) {
    return false
}
```
looks unexpected to me.  I would normally expect a function called `getSomething` to return `true` if successful and `false` otherwise.  Maybe we can rename the function to `getFiniteWellResiduals()`, make it return `true` if all residuals are finite and then write this conditional as
```C++
if (! this->getFiniteWellResiduals(...)) {
    return false;
}
```
instead?

Another option might be to extract the finite vs. non-finite residual checking out to a separate helper function and have `getWellResiduals()` just return the values as it already does.  That way we won't have to introduce an "out" parameter or add a deferred logger.  The downside to that is that we lose a little bit of context and it becomes harder to distinguish the control equation from the flow equations.

Reply:
I liked your first option. Will update accordingly.